sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Distinct.java

```diff
@@ -81,9 +86,16 @@
     return new WithRepresentativeValues<T, IdT>(fn, null);
   }
 
+  private static <T> boolean distinctAcrossPanes(PCollection<T> in) {
+    return in.getWindowingStrategy().getAllowedLateness().compareTo(Duration.ZERO) > 0
+        || !(in.getWindowingStrategy().getTrigger() instanceof DefaultTrigger);
+
+  }
   @Override
   public PCollection<T> expand(PCollection<T> in) {
-    return in
+    boolean acrossPanes = distinctAcrossPanes(in);
```

Comment:
I'd flip this and say that the two cases are the general case or the "only one pane" special case that doesn't require state.